[ZEPPELIN-1387] Support table syntax in markdown interpreter

1ambda opened the pull request:
### What is this PR for?

Support table markdown syntax issued by [ZEPPELIN-1387](https://issues.apache.org/jira/browse/ZEPPELIN-1387?jql=project%20%3D%20ZEPPELIN)

### What type of PR is it?
[Bug Fix | Improvement]

This PR can be categorized as bug fix and improvement since it not only resolves the above issue but also support other markdown syntaxes.

### Todos
* [ ] - Check the license issue of the [pegdown](https://github.com/sirthias/pegdown) library introduced by this PR

### What is the Jira issue?

[ZEPPELIN-1387](https://issues.apache.org/jira/browse/ZEPPELIN-1387?jql=project%20%3D%20ZEPPELIN)

### How should this be tested?

Write markdown texts and compare them with expected html DOMs. I'v also included some tests for this PR.

### Screenshots (if appropriate)

<img width="708" alt="markdown" src="https://cloud.githubusercontent.com/assets/4968473/18061274/1f2be526-6e5d-11e6-9f1a-3528f3958d2c.png">

### Questions:

* Does the licenses files need update?
* Does coding style is appropriate?

### Additional Comments

We might solve this issue by implementing custom table plugin for markdown4j by referring [the existing work of txtmark](https://github.com/zhenchuan/txtmark/commit/178486805e78e3d572b071ca8b9f8887a066edef). 
But I think it is not good idea in regard to coverage, maintainability and efficiency since markdown4j is currently not developed actively and it costs to implement all markdown plugins which is not supported by markdown4j.

bzz commented:
@1ambda thank you for improvement! Let me review and get back to you with it.

AhyoungRyu commented:
@1ambda Thank you for the contribution! I just shortly look through the screenshot in the description, but comparing to Github markdown support it seems Zeppelin's table doesn't have a table border. If I make the below table in Github,

```
| Tables   |      Are      |  Cool |
|----------|:-------------:|------:|
| col 1 is |  left-aligned | $1600 |
| col 2 is |    centered   |   $12 |
| col 3 is | right-aligned |    $1 |
```

it will be shown like this.

<img width="313" alt="screen shot 2016-08-30 at 11 10 51 am" src="https://cloud.githubusercontent.com/assets/10060731/18073629/8b893116-6ea2-11e6-9166-46e246acbc04.png">

But in Zeppelin it doesn't have the border.
<img width="446" alt="screen shot 2016-08-30 at 11 17 57 am" src="https://cloud.githubusercontent.com/assets/10060731/18073791/7681b1ca-6ea3-11e6-9229-3104a33332dc.png">

bzz commented:
The change of library looks great to me, 👍  for having a test case!

Few things to take care of, before merging it:
 - for all the code, please make sure it follows project [styleguide\code conventions](https://github.com/apache/zeppelin/blob/master/CONTRIBUTING.md#code-convention)
 - as you add a new dependency - it will become a part of the next release convenience binary for Zeppelin and in order to be included, we need to verify it's licence compatibility. Can you please add it to `zeppelin-distribution/src/licence_bin/LICENSE`
 - also, applying some styles for such tables, as noted by @AhyoungRyu sounds like a good improvement as well

1ambda commented:
I will refine this PR. 

Thanks for the detailed and significant review :) @AhyoungRyu @bzz

Leemoonsoo commented:
Thanks @1ambda for the improvement.

Regarding license, 'markdown4j' section in the `zeppelin-distribution/src/licence_bin/LICENSE` should be removed while the library is no longer included in the binary package.

Also `pegdown` library does have some transitive dependencies and they also need to be addressed in `zeppelin-distribution/src/licence_bin/LICENSE`. @1ambda Could you take care of them if they're not already addressed in `zeppelin-distribution/src/licence_bin/LICENSE`?

```
[INFO] +- org.pegdown:pegdown:jar:1.6.0:compile
[INFO] |  \- org.parboiled:parboiled-java:jar:1.1.7:compile
[INFO] |     +- org.parboiled:parboiled-core:jar:1.1.7:compile
[INFO] |     +- org.ow2.asm:asm:jar:5.0.3:compile
[INFO] |     +- org.ow2.asm:asm-tree:jar:5.0.3:compile
[INFO] |     +- org.ow2.asm:asm-analysis:jar:5.0.3:compile
[INFO] |     \- org.ow2.asm:asm-util:jar:5.0.3:compile
```

1ambda commented:
@Leemoonsoo Thanks!

I'v added missing transitive deps and removed unused markdown4j dep. 

I also found that `org.ow2.asm:asm:jar:3.0.1` is described in `zeppelin-distribution/src/licence_bin/LICENSE` But already denoted license is different version (3.0.1) compared to pegdown's dependency (5.0.3). 

Should I updated `org.ow2.asm:asm:jar:3.0.1` to `org.ow2.asm:asm:jar:5.0.3`?

felixcheung commented:
MArkdown4j supports YUML and Websequence, does this support the same?

1ambda commented:
@felixcheung 

Pegdown doesn't support YUML, Websequence and Include plugins according to [Pegdown Extensions](https://github.com/sirthias/pegdown/blob/master/src/main/java/org/pegdown/Extensions.java)

I think we have 2 options. (or more)

1. Use markdown4j and implement [table plugins](https://github.com/jdcasey/markdown4j/blob/master/src/main/java/org/markdown4j/TablePlugin.java) correctly. 
** we might need small effort, but we will depend on markdown4j which is not actively developed.
** and other markdown syntaxes might be implemented if necessary.

2. Use pegdown and implement [YUML](https://github.com/billy1380/markdown4j-gwt/blob/master/src/main/java/org/markdown4j/YumlPlugin.java), [Websequence](https://github.com/jdcasey/markdown4j/blob/master/src/main/java/org/markdown4j/WebSequencePlugin.java).

1ambda commented:
@AhyoungRyu  In regard to table style, 

Using CSS (e.g [github-markdown-css](https://github.com/sindresorhus/github-markdown-css))  would be better solution rather than adding inline style to DOMs such as`table`, `thead`, `tbody`, `tr`, `td` while parsing.

1ambda commented:
Now markdown table looks like

<img width="338" alt="screen shot 2016-09-01 at 3 03 02 am" src="https://cloud.githubusercontent.com/assets/4968473/18141148/54414ab6-6ff3-11e6-897d-94dab08e6516.png">

felixcheung commented:
Personally YUML and Websequence is very handy - I use them quite a bit.
What do other think about this?

AhyoungRyu commented:
@1ambda Yeah it's good idea to apply github markdown style to Zeppelin. I tested again and it works well. Definitely better than before! 

But I found little bit wired behavior when I made a title. 

```
## This is second level title
### This is third level title
```

This title code makes a link and it's connected to Zeppelin home. Is it intended?
![markdown_test](https://cloud.githubusercontent.com/assets/10060731/18164413/ad36f698-707a-11e6-91ad-3639a5f1dd7f.gif)

1ambda commented:
@AhyoungRyu 

It was due to the `ANCHORLINK` option in pegdown. I removed it. 

![markdown gif](https://cloud.githubusercontent.com/assets/4968473/18173464/a18a4d66-70a4-11e6-95b7-7d030e93cf2b.gif)

Leemoonsoo commented:
Regarding @felixcheung's comment, 

I think keeping backward compatibility is always very important.
If i add more possible options after @1ambda 's 

 - Create new interpreter %pd that uses pegdown and restore %md to use markdown4j
 - Add a interpreter property that selects markdown library pegdown/markdown4j.

AhyoungRyu commented:
@1ambda Yeah tested again and it looks good👍  Thanks!  
(One more thing: [Zeppelin's Java block indentation size](https://google.github.io/styleguide/javaguide.html#s4.2-block-indentation) is `2 spaces`  :D )

1ambda commented:
@Leemoonsoo 

I think **option 2** (Adding a property) is a better solution. because

1) Adding a new interpreter just for support another markdown parser might be confusing to users
2) Pegdown parser will be able to replace markdown4j parser soon if we implement YUML and Websequence. At that time, we can just modify the parser property of `md` interpreter instead of requesting for users to change their interpreter type

1ambda commented:
I'v restored markdown4j parser and add `markdown.parser.type` property to `md` group. Pegdown parser will be used only if user specify `pegdown` property value.

<img width="516" alt="screen shot 2016-09-03 at 1 54 12 am" src="https://cloud.githubusercontent.com/assets/4968473/18211602/66b9cf9c-7179-11e6-8506-9ae6fcfbfe3b.png">

One more question. It seems that `description` does't show even if i already provide in `markdown/src/main/resources/interpreter-setting.json` like 

```json
[
  {
    "group": "md",
    "name": "md",
    "className": "org.apache.zeppelin.markdown.MarkdownInterpreter",
    "properties": {
      "markdown.parser.type": {
        "envName": "MARKDOWN_PARSER_TYPE",
        "propertyName": "markdown.parser.type",
        "defaultValue": "markdown4j",
        "description": "Markdown Parser Type. Available values: markdown4j, pegdown. Default = markdown4j"
      }
    }
  }
]
```

AhyoungRyu commented:
@1ambda Great! It works well as expected 👍 

If you want to add the interpreter property(`markdown.parser.type`) to the markdown interpreter, it would be better to provide that information via [Zeppelin markdown interpreter docs](https://zeppelin.apache.org/docs/0.7.0-SNAPSHOT/interpreter/markdown.html) with configuration table as like the other interpreters do. Nothing much needed. You can just add the below conf table to `markdown.md`.
```
## Configuration
<table class="table-configuration">
  <tr>
    <th>Name</th>
    <th>Default Value</th>
    <th>Description</th>
  </tr>
  <tr>
    <td>markdown.parser.type</td>
    <td>markdown4j </td>
    <td>Markdown Parser Type. <br/> Available values: markdown4j, pegdown.</td>
  </tr>
</table>
```

Also the below sentences in [Overview section](https://zeppelin.apache.org/docs/0.7.0-SNAPSHOT/interpreter/markdown.html#overview) need to be updated accordingly. 
>Markdown is a plain text formatting syntax designed so that it can be converted to HTML. Apache Zeppelin uses markdown4j. For more examples and extension support, please checkout here.

FYI, `markdown.md` file is under `ZEPPELIN_HOME/docs/interpreter/` :)

1ambda commented:
@AhyoungRyu Thanks for for your review and comments :)

I'v updated docs and added missing license declaration in some java files.

Now docs/markdown looks like 

<img width="791" alt="docs-markdown-update-1" src="https://cloud.githubusercontent.com/assets/4968473/18229076/06d5b834-72a5-11e6-994c-585def006adc.png">

<img width="466" alt="docs-markdown-update-2" src="https://cloud.githubusercontent.com/assets/4968473/18229096/80ba7a22-72a5-11e6-8c8c-46a105e2bf96.png">

<img width="539" alt="docs-markdown-update-3" src="https://cloud.githubusercontent.com/assets/4968473/18229083/0b7b6bcc-72a5-11e6-9688-c1f0238634c4.png">

Leemoonsoo commented on zeppelin-web/src/app/interpreter/interpreter.css:
```diff
@@ -109,3 +109,696 @@
+
+/**
+ * github flavored markdown style
+ */
```
Are following CSS code blocks distributed under zeppelin-distribution/src/bin_license/licenses/LICENSE-github-markdown-css-2.4.0? Then the license need to added in /LICENSE and /licences directory under root and removed from bin_license. Because following codes wil be packaged into source release. 

And if we can create separate CSS file for it, that would be less confusing than different licences in a single file.

1ambda replied:
@Leemoonsoo 

I'v moved license file and description to `LICENSE` and `licenses/` from `zeppelin-distribution/src/bin_license/` 
Also, I extracted the markdown css code from `interpreter.css` to clarify its license. (297733f)

AhyoungRyu commented:
@1ambda Looks good! 👍👍  for adding screenshot images. 

>It seems that description does't show even though i provide it to markdown/src/main/resources/interpreter-setting.json like

yeah I also felt uncomfortable when create/edit a interpreter. There is no way to get information what these properties are working for. It would be better to add `description` column separately to the interpreter setting table I think. But it can be handled in the different PR :)

corneadoug commented:
@1ambda Usually to include external source css, we would use the `assets/styles` folder.

However in this case, you can use bower to have it part of the build instead of the source code, by using `bower install --save github-markdown-css`

1ambda commented:
@corneadoug Thanks! 

I'v installed github markdown css and added it to `bower.json` using the command you provided. 
It is definitely better to use bower if possible :)

corneadoug commented on docs/interpreter/markdown.md:
```diff
@@ -25,14 +25,44 @@ limitations under the License.
 In Zeppelin notebook, you can use ` %md ` in the beginning of a paragraph to invoke the Markdown interpreter and generate static html from Markdown plain text.
 
-In Zeppelin, Markdown interpreter is enabled by default.
+In Zeppelin, Markdown interpreter is enabled by default and uses [markdown4j](https://github.com/sirthias/pegdown) engine.
```
That links for markdown4j points to pegdown

1ambda replied:
@corneadoug Thanks! I missed some typos. 

They was fixed in 4d1cb3c

corneadoug commented:
@Leemoonsoo @felixcheung Are we good on that PR?
CI is failing on tests, not related to this PR as its happening to other PRs.
But would be better to merge this one after CI is fixed

felixcheung commented:
looks fine to me, I'd defer to others on licensing. thanks for the great contribution!

Leemoonsoo commented:
Because now github-markdown-css is downloaded on build time instead of packed in the source tree, related license now need to moved back to under bin_licenses.

bzz commented on markdown/src/main/java/org/apache/zeppelin/markdown/MarkdownInterpreter.java:
```diff
@@ -17,50 +17,83 @@
-    } catch (IOException | java.lang.RuntimeException e) {
-      LOGGER.error("Exception in Markdown while interpret ", e);
+      html = parser.render(markdownText);
+    } catch (Exception e) {
```
Is there a reason to catch general Exception instead of more specific sub-classes?

bzz commented:
@1ambda 👍  fot tests, looks great to me except for licensing issue raised by Moon and small comment above.

If you could let me know you user name in ASF JIRA I'll be happy to help assigning ZEPPELIN-1387 to you.

1ambda commented:
@Leemoonsoo I'v fixed license description. Also, now I understand when to describe licenses to `licenses` or `bin_licenses`. Thanks. :)

@bzz I think it is better to use more specific exception types in catch block as you pointed out. Could you review e08929a again?

Leemoonsoo commented:
Thanks @1ambda. Looks great to me.
Merge if there're no further discussions.

AhyoungRyu commented:
@1ambda Sadly the conflict needs to be resolved first before merging. But only for `docs/interpreter/markdown.md` this file :)

1ambda commented:
@AhyoungRyu 

I'v resolved the merge conflict with 8f344db. Could you try again? :)

1ambda commented:
<img width="1094" alt="screen shot 2016-09-21 at 1 39 41 am" src="https://cloud.githubusercontent.com/assets/4968473/18679870/5ae08dec-7f9c-11e6-8c61-91e87bad5ed9.png">

CI is failing on only the last job.

AhyoungRyu commented:
@1ambda Yeah tested locally and looks good. Thanks for your quick response! Now it's really ready to be merged I think 👍 
The other PRs' CI are failing in Selenium test because of the same reason. Currently it's tracked in [ZEPPELIN-1449](https://issues.apache.org/jira/browse/ZEPPELIN-1449).

corneadoug commented:
That CI failure is being handled in #1444, you should be able to rebase and have a green CI soon

corneadoug commented:
@1ambda You can rebase now :)

1ambda commented:
Now CI is failing on the third build job. :( 

I checked out `zeppelin/master` and merged it on `1ambda/fix-zeppelin-1387`. 
Should I have use git `rebase` command instead? I thought using `rebase` command will overwrite 
this PR history (comments, etc...)

<img width="1080" alt="screen shot 2016-09-21 at 11 00 10 am" src="https://cloud.githubusercontent.com/assets/4968473/18695206/2afb0c0c-7feb-11e6-8d94-119a456876a0.png">
<img width="1100" alt="screen shot 2016-09-21 at 11 00 26 am" src="https://cloud.githubusercontent.com/assets/4968473/18695207/2cd2ce8e-7feb-11e6-961d-5292d58513cf.png">

1ambda commented:
Close PR to re-trigger CI

1ambda commented:
Open PR to re-trigger CI

corneadoug commented:
Usually we prefer rebase since it doesn't pollute the commit history. But its fine, the end result is the same. We sometimes have flaky tests, I will check once the build is finished

1ambda commented:
Finally, CI is green! :)

corneadoug commented:
Awesome, Merging if there is no more discussions
